### PR TITLE
#106 Dashboard Refactoring and Database Dependency inversion

### LINKED ISSUE #105: Refactoring dash board
As a DEVELOPER, I want to refactor dashboard and utils to follow the dependency inversion principle, and ensure that the Heroku instructions are correctly stated in the app.json file.

A/C:

- [ ] Use_cases in dashboard depend (high level) depend on abstract_data_access class, and the firebase_data_access_implementation (low level) will inherit from abstract_data_access class. This way, if we were to add another data base than firebase, we can easily add another class and inherit it from abstract_data_access class
- [ ] Ensure that Heroku commands are running standard-2x instead of performance M, app.json should be in the root directory.

### PR DESCRIPTION
**Pull request by Chong-source**
1. Use_cases in dashboard depend (high level) depend on abstract_data_access class, and the firebase_data_access_implementation (low level) will inherit from abstract_data_access class. This way, if we were to add another data base than firebase, we can easily add another class and inherit it from abstract_data_access class

Note: I have moved abstract_data_access and firebase data access classes to the utils folder. Any files that require data access is now using `from utils.abstract_data_access import AbstractDataAccess`

2. Ensure that Heroku commands are running standard-2x instead of performance M, app.json should be in the root directory.

Note: I was having trouble modifying the instructions for heroku until Jennifer pointed out that my app.json was in the backend directory, now it's moved to the root directory.

Resolves #105

### CODE REVIEW

File: backend/dashboard/use_cases.py [hunk 1]
```diff
@@ -1,128 +1,122 @@
...
-    """
-    Returns the difference between last month and this month's # of green transactions.
-    """
-    user_transactions = get_table_from_firebase('Users')[user_id]['transactions']
-    esg_data = get_table_from_firebase('esg')
-    monthly_green_transactions = calculate_historical_green_transactions(user_transactions, esg_data)
-    if monthly_green_transactions[0] is None or monthly_green_transactions[1] is None:
-        return 0
-    return int(monthly_green_transactions[0] - monthly_green_transactions[1])
+class DashboardUseCases:
```
Comment (YoyoLiuuu) [inline backend/dashboard/use_cases.py:6]: Perhaps some comments explaining what this function does?

File: backend/dashboard/calculations.py [hunk 2]
```diff
@@ -3,211 +3,212 @@
...
-
-def _get_unique_companies(transactions: list[dict], ESG_scores: dict[str, dict]):
-    """
-    Return set of top companies shopped at for all time
-    """
-    unique_companies = set()
-
-    for transaction in transactions:
+class Calculations:
+    def _get_closest_match(query: str, choices: dict, score_cutoff: int = 75) -> str:
```
Comment (jnnchi) [inline backend/dashboard/calculations.py:7]: I think you need the self parameter in these functions, or need to label them as static

File: backend/dashboard/views.py [hunk 3]
```diff
@@ -1,70 +1,62 @@
...
-# increase/decrease of CO2 score
-def get_co2_score_change(request): 
-    user_id = request.session.get("user_id") or '0'
-    return JsonResponse(co2_score_change(user_id), safe=False)
-
-# percent increase/decrease of green transactions
-def get_green_transaction_change(request): 
-    user_id = request.session.get("user_id") or '0'
-    return JsonResponse(green_transaction_change(user_id), safe=False)
+class DashboardView:
```
Comment (YoyoLiuuu) [inline backend/dashboard/views.py:7]: also would be helpful with comments!

File: backend/utils/abstract_data_access.py [hunk 4]
```diff
@@ -0,0 +1,7 @@
+from abc import ABC, abstractmethod
+
+#
+class AbstractDataAccess(ABC):
```
Comment (YoyoLiuuu) [inline backend/utils/abstract_data_access.py:4]: here too!

File: backend/dashboard/calculations.py [hunk 5]
```diff
@@ -3,211 +3,212 @@
...
+
+
+    def _get_unique_companies(transactions: list[dict], ESG_scores: dict[str, dict]):
+        """
+        Return set of top companies shopped at for all time
+        """
+        unique_companies = set()
+
+        for transaction in transactions:
+            company_name = _get_closest_match(transaction['merchant_name'], ESG_scores)
```
Comment (jnnchi) [inline backend/dashboard/calculations.py:79]: These helper functions aren't recognized now that they're within a class. I think you need to add the self parameter into each other methods, and then call them using self._get_closest_match

File: backend/utils/firebase_data_access_implementation.py [hunk 6]
```diff
@@ -0,0 +1,16 @@
+from utils.firebase import db
+from .abstract_data_access import AbstractDataAccess
+class FirebaseDataAccess(AbstractDataAccess):
```
Comment (YoyoLiuuu) [inline backend/utils/firebase_data_access_implementation.py:3]: class comment here

File: backend/utils/firebase_data_access_implementation.py [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+    def get_table_from_database(table_to_access: str) -> dict:
```
Comment (jnnchi) [inline backend/utils/firebase_data_access_implementation.py:4]: Needs self parameter

**Review by YoyoLiuuu — COMMENTED** [review]
Everything looks good! @Chong-source It would be great if you can add some comment to explain these classes.

File: backend/esg/views.py [hunk 8]
```diff
@@ -6,52 +6,53 @@
...
-    with open(esg_path, 'r') as file:
-        data = json.load(file)
-    
-    try:
-        doc = db.collection('test').document('testDoc')
-        doc.set({'connected': True})
-        print("Firebase is connected, and data was written successfully.")
-    except Exception as e:
-        print("Connection Error:", e)
+    def upload_data_to_firestore(request):
```
Comment (jnnchi) [inline backend/esg/views.py:11]: Also need self parameter

File: backend/dashboard/views.py [hunk 9]
```diff
@@ -1,70 +1,62 @@
...
+    
+    def __init__(self, use_cases: DashboardUseCases):
+        self.use_cases = use_cases
+            
+    # monthly carbon score, green transactions, and month names for each month - return the last 12 data points as a list
+    def get_line_graph_data(request):
```
Comment (jnnchi) [inline backend/dashboard/views.py:13]: I think these also need self parameter

**Review by jnnchi — CHANGES_REQUESTED** [review]
Great work, thank you Chong! One thing is that I think the self parameter needs to be added to all class methods, unless you declare them as static. Helper method calls also need to be called with self._helper_method() instead of just _helper_method()

File: backend/dashboard/calculations.py (hunk 2 above)
Comment (Chong-source) [reply backend/dashboard/calculations.py:7]: Thank you Jennifer for catching that! I fixed it.

File: backend/dashboard/use_cases.py [hunk 10]
```diff
@@ -1,128 +1,124 @@
...
+    """DashboardUseCases class is responsible for handling the business logic of the dashboard.
+    """
+    def __init__(self, calculations: Calculations, data_access: AbstractDataAccess):
+        self.calculations = calculations
+        self.data_access = data_access
+    def past_12_month_names() -> list[str]:
```
Comment (jnnchi) [inline backend/dashboard/use_cases.py:12]: Can the self parameter also be added to the functions in this file?

**Review by jnnchi — CHANGES_REQUESTED** [review]

File: backend/dashboard/use_cases.py (hunk 10 above)
Comment (Chong-source) [reply backend/dashboard/use_cases.py:12]: Yes, that's my bad I overlooked that

**Review by jnnchi — APPROVED** [review]

**Comment by jnnchi** [comment]
Looks great, thank you!

**Comment by benrockehenderson** [comment]
This looks good to go!

**Pull request merged**